### PR TITLE
CXP-415: Add the type of the author to the webhook event

### PR DESCRIPTION
**Pull request by pierrejolly**
Add the type of the author to the webhook event (`ui` or `api`)

### CODE REVIEW

File: src/Akeneo/Platform/Component/EventQueue/BusinessEvent.php [hunk 1]
```diff
@@ -13,17 +13,20 @@
...
 {
     private $author;
+    private $authorType;
     private $data;
     private $timestamp;
     private $uuid;
 
     public function __construct(
         string $author,
+        string $authorType,
```
Comment (pchasle) [inline src/Akeneo/Platform/Component/EventQueue/BusinessEvent.php:23]: Maybe we should use a VO Author. What do you think?

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductBusinessEventSubscriber.php [hunk 2]
```diff
@@ -53,13 +53,14 @@ public function produceBusinessSaveEvent(GenericEvent $event): void
         }
 
         $author = $user->getUsername();
+        $authorType = $user->getType();
```
Comment (pchasle) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductBusinessEventSubscriber.php:56]: From what I see `User::$type` will contains `api` or `user` (not `ui`). This won't work.
What I suggest would be to add a VO `Author` with a factory method `Author::fromUser(UserInterface $user): Author`

Comment (pchasle) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductBusinessEventSubscriber.php:56]: That will allow you to manage easily our type of user `api` or `ui` using the the `User::isApiUser` method

**Comment by pierrejolly** [comment]
GTM 👍

**Review by tomglvng — APPROVED** [review]
Approved by Pierre.

File: src/Akeneo/Platform/Component/EventQueue/BusinessEventNormalizer.php [hunk 3]
```diff
@@ -52,8 +63,16 @@ public function denormalize($data, $type, $format = null, array $context = [])
             throw new RuntimeException(sprintf('The class "%s" is not defined.', $type));
         }
 
+        $user = $this->userRepository->findOneByIdentifier($data['author']);
+
+        if (!$user instanceof UserInterface) {
+            throw new \LogicException('User not found.');
+        }
+
+        $author = Author::fromUser($user);
```
Comment (pchasle) [inline src/Akeneo/Platform/Component/EventQueue/BusinessEventNormalizer.php:72]: No need for the repository here, you need to build the Author VO from the data you already have:
`new Author($data['author'], $data['author_type'])`

File: src/Akeneo/Platform/Component/EventQueue/Author.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    /** @var string */
+    private $name;
+
+    /** @var string */
+    private $type;
+
+    private function __construct(string $name, string $type)
+    {
+        $this->name = $name;
+        $this->type = $type;
```
Comment (pchasle) [inline src/Akeneo/Platform/Component/EventQueue/Author.php:30]: You need to validate that the type is valid

File: src/Akeneo/Tool/Bundle/ApiBundle/tests/integration/ApiTestCase.php [hunk 5]
```diff
@@ -303,43 +307,32 @@ protected function executeStreamRequest(
...
+            self::USERNAME,
+            self::PASSWORD,
+            'admin@example.com',
+            'John',
+            'Doe'
+        );
+    }
 
-        return $user;
+    protected function createUser(string $username): UserInterface
```
Comment (pchasle) [inline src/Akeneo/Tool/Bundle/ApiBundle/tests/integration/ApiTestCase.php:327]: Why the change? This function doesn't seems to be used?!

File: src/Akeneo/Platform/Component/EventQueue/Author.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    /** @var string */
+    const TYPE_UI = 'ui';
+
+    /** @var string */
+    private $name;
+
+    /** @var string */
+    private $type;
+
+    private function __construct(string $name, string $type)
```
Comment (pchasle) [inline src/Akeneo/Platform/Component/EventQueue/Author.php:27]: The constructor should be public. Having a factory method to help you build an object doesn't means you can't use the "classic" way to build it.

File: tests/back/Pim/Enrichment/Specification/Component/Product/Message/ProductCreatedSpec.php [hunk 7]
```diff
@@ -14,10 +16,17 @@
...
 {
-    public function let(): void
+    public function let(UserInterface $user): void
     {
+        $user->getUsername()->willReturn('julia');
+        $user->getFirstName()->willReturn('Julia');
+        $user->getLastName()->willReturn('Doe');
+        $user->isApiUser()->willReturn(true);
+
+        $author = Author::fromUser($user->getWrappedObject());
```
Comment (pchasle) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Message/ProductCreatedSpec.php:26]: You can simplify a lot of your tests by using the constructor `new Author('julia', 'ui')`

File: src/Akeneo/Tool/Bundle/ApiBundle/tests/integration/ApiTestCase.php (hunk 5 above)
Comment (tomglvng) [reply src/Akeneo/Tool/Bundle/ApiBundle/tests/integration/ApiTestCase.php:327]: When I was using user repository, I had to create a non-admin user. I let this function just in case someone needs it. Should I remove it ?

Comment (tomglvng) [reply src/Akeneo/Tool/Bundle/ApiBundle/tests/integration/ApiTestCase.php]: Saw together : remove it.

File: src/Akeneo/Platform/Component/EventQueue/Author.php (hunk 6 above)
Comment (tomglvng) [reply src/Akeneo/Platform/Component/EventQueue/Author.php:28]: Saw together: I let it private.

File: tests/back/Platform/Specification/Component/EventQueue/BusinessEventNormalizerSpec.php [hunk 8]
```diff
@@ -96,10 +121,45 @@ public function it_does_not_supports_denormalization_of_non_business_event(): vo
...
+            'timestamp' => 0,
+            'uuid' => 'e0e4c95d-9646-40d7-be2b-d9b14fc0c6ba',
+        ];
+
+        $denormalizedBusinessEvent = $this->denormalize($data, get_class($businessEvent));
+
+        Assert::assertEquals($denormalizedBusinessEvent->getWrappedObject(), $businessEvent);
+    }
+
+    public function it_does_not_denormalize_a_business_event_because_author_type_is_not_good(UserInterface $user)
```
Comment (pchasle) [inline tests/back/Platform/Specification/Component/EventQueue/BusinessEventNormalizerSpec.php:154]: Just a minor preference over Iznogoud
```suggestion
    public function it_does_not_denormalize_a_business_event_because_author_type_is_invalid(UserInterface $user)
```

**Review by pchasle — APPROVED** [review]
